src/validation.cairo

```diff
@@ -37,10 +39,56 @@ impl BlockValidatorImpl of BlockValidator {
 
 #[generate_trait]
 impl TransactionValidatorImpl of TransactionValidator {
+    // marker, flag, and witness fields in segwit transactions are not included
+    // this means txid computation is the same for legacy and segwit tx
     fn txid(self: @Transaction) -> u256 {
-        // TODO: implement
-        0
+        let version: ByteArray = (*self.version).format_as_byte_array(10);
+        let inputs: Span<TxIn> = *self.inputs;
+        let outputs: Span<TxOut> = *self.outputs;
+        let locktime: ByteArray = (*self.lock_time).format_as_byte_array(10);
+
+        // append version
+        let mut sha256_input: ByteArray = "";
+        sha256_input.append(@version);
+
+        // append inputs
+        let mut i = 0;
+        while i < inputs.len() {
+            sha256_input.append(*inputs.at(i).script);
```

Comment:
Maybe it makes sense to introduce a "serializable" trait and implement it for inputs/outputs/transaction? And also for Span<T: Serializable>
That way it would be much clearer

Reply:
It should serialize to Array(u32)